### DISCUSSION TITLE
Script support

### DISCUSSION THREAD

**Discussion by violinyanev** [google/filament#5029] [Ideas]
Hi there!

TLDR: are there any plans/desire to add scripting support to Filament (similar to Blueprints in Unreal or the C# subsystem in Unity)?

For the past ~10 years I've been creating, maitaining and finally open-sourcing the rendering system we use in my company. Today, we are still maintaining our own rendering engine, even though there are good open source alternatives with a live community (Filament being the primary option for us). But why? Two reasons mostly:
- none of the OSS rendering engines (Ogre, O3DE, Godot) has embedded platforms in its focus (sure, they run on them, but more for compatibility then for getting the maximum out of them). This is also the case for commercial engines like UE or Unity btw...
- Filament and glTF are exactly what an embedded platform needs to run 3D gfx smoothly, but doesn't support scripting yet (despite my hope it will just magically appear at some point because all other engines have it)

Here is a link to the scripting runtime we use for our rendering engine - in case you are interested to understand more closely what we need and some rationale why:
https://ramses-logic.readthedocs.io/en/latest/readme_ref.html#introduction

I am curious to your feedback!

**Comment by romainguy** [comment]
Thanks for the suggestion. Filament aims to be just a renderer. A scripting runtime is definitely out of scope but could be provided as a separate project independently.

**Comment by violinyanev** [comment]
Hi! Thanks for the quick reply! Understood.

In case you want to reconsider some time later - here is a summary of what we found out. We did what you suggest (split the script/animation runtime as a separate project). Works quite well initially. In retrospective, I think it would have been better to keep them together though. Things like animation for example are not that well attributable to pure rendering or pure scripting - the keyframe data is somehow aligned with rendering attributes (pose, material properties etc), but also has tight connection to state and logic (running, paused, time units). I think that's why most commercial engines include animation and scripting and integrate them tightly. We are starting to learn this too.
